File: 백준/Gold/16234. 인구 이동/인구 이동.py
from collections import deque
import sys
input = sys.stdin.readline

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

def bfs(graph, s_x, s_y, visited):
  population = 0
  queue = deque()
  queue.append((s_x, s_y))

  union = []
  union.append((s_x, s_y))

  visited[s_x][s_y] = True

  while queue:
    x, y = queue.popleft()

    for i in range(4):
      nx = x + dx[i]
      ny = y + dy[i]
      
      if nx < 0 or ny < 0 or nx >= n or ny >= n:
        continue
      if visited[nx][ny] == False and (l <= abs(graph[nx][ny] - graph[x][y]) <= r):

        visited[nx][ny] = True
        queue.append((nx, ny))
        union.append((nx, ny))

  if len(union) <= 1:
    return 0
  else:
    for x, y in union:
      population += graph[x][y]
      
    population = int(population / len(union))

    for x, y in union:
      graph[x][y] = population

    return 1



n, l, r = map(int, input().split())
land = [[0 for _ in range(n)] for _ in range(n)]
time = 0

for i in range(n):
  land[i] = list(map(int, input().split()))

visited = [[False for _ in range(n)] for _ in range(n)]

while 1:
  flag = 0
  visited = [[False for _ in range(n)] for _ in range(n)]

  for i in range(n):
    for j in range(n):
      if visited[i][j] == False:
        flag += bfs(land, i, j, visited)
        
  if flag == 0:
    break
  
  time += 1
  
print(time)

File: 백준/Gold/16234. 인구 이동/인구 이동.cc
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>
#include <list>
#include <string>
#include <hash_map>
#include <unordered_map>
using namespace std;

#define endl '\n'

int dx[4] = {-1, 0, 1, 0};
int dy[4] = {0, 1, 0, -1};

int n, l, r, ans;
int land[51][51];

int bfs(pair<int, int> start, bool visited[51][51]) {
    queue<pair<int, int> > q;
    q.emplace(start.first, start.second);

    vector<pair<int, int> > country;
    country.emplace_back(start.first, start.second);

    visited[start.first][start.second] = true;

    while (!q.empty()) {
        const auto cur = q.front();
        q.pop();

        for (int i = 0; i < 4; ++i) {
            int nx = cur.first + dx[i];
            int ny = cur.second + dy[i];

            if (nx < 0 || ny < 0 || nx >= n || ny >= n) {
                continue;
            }

            if (visited[nx][ny] == false && (l <= abs(land[nx][ny] - land[cur.first][cur.second]) && abs(land[nx][ny] - land[cur.first][cur.second]) <= r)) {
                visited[nx][ny] = true;
                q.emplace(nx, ny);
                country.emplace_back(nx, ny);
            }
        }
    }

    if (country.size() <= 1) {
        return 0;
    }
    int population = 0;
    for (const auto pos: country) {
        population += land[pos.first][pos.second];
    }

    population = population / country.size();

    for (const auto pos: country) {
        land[pos.first][pos.second] = population;
    }

    return 1;
}


int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    cin >> n >> l >> r;

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            cin >> land[i][j];
        }
    }

    while (true) {
        int flag = 0;
        bool visited[51][51] = {false, };

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (!visited[i][j]) {
                    flag += bfs({i, j}, visited);
                }
            }
        }

        if (!flag) {
            break;
        }

        ans++;
    }

    cout << ans;
}
